File: 7-grafos/1135.c
#include<stdlib.h>
#include<stdio.h>
#include<math.h>
//

typedef struct Vector{
	int*array,index;
	size_t length;
}Vector;

struct Vector*createVector(size_t size){
	Vector*newVector=(Vector*)malloc(sizeof(Vector));
	newVector->array=(int*)malloc(size*sizeof(int));
	newVector->index=0;
	newVector->length=size;
	return newVector;
}
void appendElementVector(Vector*ref,int value){
	if((size_t)ref->index>ref->length-1){
		ref->array=realloc(ref->array,ref->length*2*sizeof(int));
		ref->length*=2;
	}
	ref->array[ref->index]=value;
	++ref->index;
}

typedef struct AdjList{
	Vector**list;
	size_t length;
}AdjList;

struct AdjList*createAdj(size_t AmountV,size_t AmountItemsV){
	AdjList*newAdj=(AdjList*)malloc(sizeof(AdjList));
	newAdj->list=(Vector**)malloc(AmountV*sizeof(Vector*));
	for(int i=0;i<(int)AmountV;++i)newAdj->list[i]=createVector(AmountItemsV);
	newAdj->length=AmountV;
	return newAdj;
}

AdjList*adj_list;
Vector*pathDFS;
int*inAdj,*indexs;
unsigned long long*weight;

void DFS(int node){
	if(!inAdj[node])return;
	for(int i=0;i<adj_list->list[inAdj[node]-1]->index;++i){
		int vertic=adj_list->list[inAdj[node]-1]->array[i];
		appendElementVector(pathDFS,node);
		(!indexs[node])?indexs[node]=pathDFS->index:indexs[node];
		DFS(vertic);
		appendElementVector(pathDFS,vertic);
		(!indexs[vertic])?indexs[vertic]=pathDFS->index:indexs[vertic];
	}
}
int logTw(int n){
	int i;
	for(i=1;(1<<(i+1))<n+1;++i){}
	return i;
}
int**logTwo(Vector*aray,int amountV){
	int**result=(int**)malloc(amountV*sizeof(int*)),logg=logTw(amountV)+1;
	for(int i=0;i<amountV;++i){
		result[i]=(int*)malloc((logg*sizeof(int)));
		result[i][0]=i;
	}
	for(int i=1;(1<<i)<amountV+1;++i){
		for(int j=0;(j+(1<<i)-1)<amountV;++j){
			if(aray->array[result[j][i-1]]<aray->array[result[j+(1<<(i-1))][i-1]])result[j][i]=result[j][i-1];
			else result[j][i]=result[j+(1<<(i-1))][i-1];
		}
	}
	return result;
}

void ProcessQ(int amountQ,int amountV){
	indexs=(int*)calloc(amountV,sizeof(int));
	pathDFS=createVector((amountV-1)*2);
	DFS(0);
	free(adj_list);
	free(inAdj);
	int**min=logTwo(pathDFS,pathDFS->index),logg=logTw(pathDFS->index)+1;
	for(int i=0;i<amountQ;++i){
		int source,destiny,ind1,ind2,tempLog;
		scanf("%i %i",&source,&destiny);
		if(indexs[source]>indexs[destiny]){
			destiny+=source;
			source=destiny-source;
			destiny-=source;
		}
		//	
		ind1=indexs[source]-1;
		ind2=indexs[destiny]-1;
		tempLog=logTw(ind2-ind1+1);
		int commonVertic=(pathDFS->array[min[ind1][tempLog]]<pathDFS->array[min[ind2-(1<<tempLog)+1][tempLog]]+1)?pathDFS->array[min[ind1][tempLog]]:pathDFS->array[min[ind2-(1<<tempLog)+1][tempLog]];
		printf("%llu",weight[source]+weight[destiny]-weight[commonVertic]*2);
		if(i<amountQ-1)printf(" ");
		else printf("\n");
	}
	free(pathDFS);
	free(indexs);
	free(weight);
}
//
int main(){
	int rep;
	while(scanf("%i",&rep)&&rep){
		int*tempRep=(int*)malloc(rep*sizeof(int)),sizeAdj=0;
		inAdj=(int*)calloc(rep,sizeof(int));
		weight=(unsigned long long*)calloc(rep,sizeof(unsigned long long));
		//
		for(int i=1;i<rep;++i){
			unsigned long long length;
			int v;
			scanf("%i %llu",&v,&length);
			if(!inAdj[v]){
				++sizeAdj;
				inAdj[v]=sizeAdj;
			}
			tempRep[i-1]=v;
			weight[i]=weight[v]+length;
		}
		adj_list=createAdj(sizeAdj,rep/sizeAdj);
		for(int i=0;i<rep-1;++i)appendElementVector(adj_list->list[inAdj[tempRep[i]]-1],i+1);
		int amountQueries;
		scanf("%i",&amountQueries);
		ProcessQ(amountQueries,rep);
	}
	return 0;
}

File: 7-grafos/1135.py
import sys

adjList,inAdj,weight,pathDFS,index=None,None,None,[],[]

def DFS(node,size)->None:
    stack=[node]
    lastNode=[0]
    explored=[0 for i in range(size)]
    while stack:
        pathDFS.append(stack[-1]);
        index[stack[-1]]=len(pathDFS) if not index[stack[-1]] else index[stack[-1]]
        if not inAdj[stack[-1]]:
            stack.pop()
            lastNode.pop()
            continue
        for i in range(lastNode[-1],len(adjList[inAdj[stack[-1]]-1])):
            aux=adjList[inAdj[stack[-1]]-1][i]
            lastNode[-1]+=1
            stack.append(aux)
            stack.append(-1)
            lastNode.append(0)
            lastNode.append(-1)
            explored[aux]=1
            break
        stack.pop()
        lastNode.pop()

def log2(n)->int:
    i=1
    while((1<<(i+1))<n+1):
        i+=1
    return i

def logTw2(size):
    result=[[0 for i in range(log2(size)+1)]for i in range(size)]
    for i in range(size):result[i][0]=i
    i=1
    while((1<<i)<size+1):
        j=0
        while(j+(1<<i)-1<size):
            result[j][i]=result[j][i-1]if pathDFS[result[j][i-1]]<pathDFS[result[j+(1<<(i-1))][i-1]] else result[j+(1<<(i-1))][i-1]
            j+=1
        i+=1
    return result
def ProcessQ(amountQ,amountV)->None:
    for i in range(amountV):
        index.append(0)
    DFS(0,amountV);
    adjList.clear()
    inAdj.clear()
    small=logTw2(len(pathDFS))
    for i in range(amountQ):
        [source,destiny]=(sys.stdin.readline().strip()).split()
        source=int(source)
        destiny=int(destiny)
        if index[source]>index[destiny]:
            destiny+=source
            source=destiny-source
            destiny-=source
        ind1,ind2=index[source]-1,index[destiny]-1
        temp=log2(ind2-ind1+1)
        commonVertic=pathDFS[small[ind1][temp]] if pathDFS[small[ind1][temp]]<pathDFS[small[ind2-(1<<temp)+1][temp]] else pathDFS[small[ind2-(1<<temp)+1][temp]]
        sys.stdout.write(str(weight[source]+weight[destiny]-weight[commonVertic]*2))
        if(i<amountQ-1):sys.stdout.write(' ')
        else: sys.stdout.write('\n')
    pathDFS.clear()
    index.clear()

while (x:=int(input())):
    inAdj,tempRep,sizeAdj=[0 for i in range(x)],[],0
    weight=inAdj.copy()
    for i in range(1,x):
        [v,length]=(sys.stdin.readline().strip()).split()
        if not inAdj[int(v)]:
            sizeAdj+=1
            inAdj[int(v)]=sizeAdj
        weight[i]=weight[int(v)]+int(length)
        tempRep.append(int(v))
    adjList=[[] for i in range(sizeAdj)]
    for i in range(x-1):
        adjList[inAdj[tempRep[i]]-1].append(i+1)
    amountQ=int(input())
    ProcessQ(amountQ,x)
